Rename AnalyzeRequest to AnalyzeTextOptions

heaths opened the pull request:
Fixes #12486

heaths committed:
Rename AnalyzeResult to AnalyzeTextResult

Fixes #12486

tg-msft commented on sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:
```diff
@@ -212,9 +212,9 @@ public partial class AnalyzedTokenInfo
         public string Token { get { throw null; } }
     }
-    public partial class AnalyzeRequest
+    public partial class AnalyzeTextRequest
```
Why not AnalyzeTextOptions?  I thought that's where we landed in the meeting.

heaths replied:
See my thread in our Teams channel. Basically, nothing about it is optional. According to [guidelines](https://azure.github.io/azure-sdk/general_design.html#client-interface):
> There are times when operations require the addition of optional data, provided in what is colloquially known as an “options bag”. Libraries should strive for consistent naming.
But `Text` is not optional, and one of `Analyzer` or `Tokenizer` is required by the service. The other option is, then, to have overloads requiring one or the other, but:
1. Not all languages can do that, and the type (string) would be the same anyway, and
2. That sounds like client-side validation of non-path arguments, which is against [guidelines](https://azure.github.io/azure-sdk/general_implementation.html#parameter-validation).
> DO NOT validate service parameters. This includes null checks, empty strings, and other common validating conditions. Let the service validate any request parameters.
If you or @KrzysztofCwalina have a different interpretation or a better suggestion, I'm happy to make a change.

heaths replied:
/cc @sima-zhu @alzimmermsft @xiangyan99 @sarangan12 @xirzec

tg-msft replied:
Here's the C# specific guidance:  https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-parameters  (The example code has a bug where `BlobCreateOptions` doesn't have `string BlobName` but Krzysztof is going to fix that.)

We had a meeting earlier this week to clarify it a bit in relation to Java.  My current understanding is that `*Options` doesn't mean optional.  Here's what I'd expect now:

```C#
AnalyzeText(string text, Analyzer analyzer);
AnalyzeText(string text, Tokenizer tokenizer);
AnalyzeText(AnalyzeTextOptions options);

public class AnalyzeTextOptions
{
    public AnalyzeTextOptions(string text, Analyzer analyzer);
    public AnalyzeTextOptions(string text, Tokenizer tokenizer);
    public string Text { get; }
    public Analyzer Analyzer { get; }
    public Tokenizer Tokenizer { get; }
    public object WhateverElseMutable { get; set; }
}
```

I think it's too late to change this consistently everywhere, but I see no harm in calling the type `AnalyzeTextOptions`.  I don't feel strongly though so I'll defer to Krzysztof.

sima-zhu replied:
Java has made the change to AnalyzeTextRequest. 
@JonathanGiles could you suggest the best naming for Java?

sima-zhu replied:
[Guidance ](https://azure.github.io/azure-sdk/java_design.html#model-types)for java

heaths replied:
Java is a bit looser in their description:
> Optional parameters and settings to an operation should be collected into an options bag named <operation>Options

As such, I'm fine renaming this to `AnalyzeTextOptions` but not the overloads, which is too much like client-side validation of model properties. The swagger lists only `text` as required, though it probably should've used `oneOf` for the others (though, I'm not sure generators support that - just `required`). We can always add it later if we see a lot of confusion by customers.

sima-zhu replied:
https://github.com/Azure/azure-sdk-for-java/pull/11783

KrzysztofCwalina replied:
"Options" in the context of the option bag parameter names does not mean "optional". It's more like options on the fix prix menu, where you need to pick at least one option or you will end up hungry, yet with ah check to pay.

heaths replied:
@KrzysztofCwalina should we also have overloads (either method and/or model constructor overloads) that take either an `Analyzer` or `Tokenizer`? That really trends toward client-side validation of properties that aren't listed in swagger's `required`, but could be helpful. If so, I can get bugs opened across languages immediately.

heaths replied:
I'm going to send out an email to clarify our intent and open bugs across languages.

heaths committed:
Rename AnalyzeRequest to AnalyzeTextOptions

Fixes #12530

tg-msft commented on sdk/search/Azure.Search.Documents/CHANGELOG.md:
```diff
@@ -6,7 +6,7 @@
 - Moved models for managing indexes, indexers, and skillsets to `Azure.Search.Documents.Indexes.Models`.
 - Split `SearchServiceClient` into `SearchIndexClient` for managing indexes, and `SearchIndexerClient` for managing indexers, both of which are now in `Azure.Search.Documents.Indexes`.
-- Renamed `AnalyzeRequest` to `AnalyzeTextRequest`.
+- Renamed `AnalyzeRequest` to `AnalyzeTextOptions`, and overloaded constructors to make constructing it 
```
Nit - forgot the end of this

heaths committed:
Rename some properties, parameters for consistency

Fixes #12536

heaths committed:
Removed the SynonymMap.Format property

Only "solr" is supported, and this makes .NET consistent with the other languages.

tg-msft commented on sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:
```diff
@@ -109,8 +109,8 @@ public partial class SearchIndexClient
         public virtual string ServiceName { get { throw null; } }
-        public virtual Azure.Response<System.Collections.Generic.IReadOnlyList<Azure.Search.Documents.Indexes.Models.AnalyzedTokenInfo>> AnalyzeText(string indexName, Azure.Search.Documents.Indexes.Models.AnalyzeRequest analyzeRequest, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
-        public virtual System.Threading.Tasks.Task<Azure.Response<System.Collections.Generic.IReadOnlyList<Azure.Search.Documents.Indexes.Models.AnalyzedTokenInfo>>> AnalyzeTextAsync(string indexName, Azure.Search.Documents.Indexes.Models.AnalyzeRequest analyzeRequest, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+        public virtual Azure.Response<System.Collections.Generic.IReadOnlyList<Azure.Search.Documents.Indexes.Models.AnalyzedTokenInfo>> AnalyzeText(string indexName, Azure.Search.Documents.Indexes.Models.AnalyzeTextOptions request, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
```
Nit - these param names should probably be `options`

heaths replied:
Huh. I could swear I changed those I wonder if that was another incarnation that got reverted. Thanks, though. I'll get that changed.

heaths committed:
Resolve PR feedback

heaths commented:
/azp run net - search - ci
